### PR TITLE
Handle GPS

### PR DESCRIPTION
**Pull request by dave-mills**
Adds basic handling of geopoints into custom data models from ODK forms. 

This also fixes a potential bug: previously, all $class entries with the current $submissionId were deleted at the end of every foeach loop, meaning that only the final entry from each repeat persisted in the table. 

This PR moves that call to before the foreach loop, so the deletion only happens once. 

@dan-tang-ssd - this might be just for you to take the GPS code into your work. As you're refactoring the functions a bit first, this might become old very quickly.

### CODE REVIEW

**Comment by dan-tang-ssd** [comment]
This PR is now ready for review.

New farms will be created if farm_id is -99.

In table main_surveys, I changed farm_id as a foreign key.
When handling main survey or repeat group, if it is mapped to a dataset with database table.
It will also get foreign key column name and foreign key table name from that table.
New record will be created in foreign key table. And then newly created record ID will be used as farm_id in main_surveys.

P.S. New farm creation business logic should belong to application instead of package. We need to think about how to move this logic from package to application. E.g. Subclass to overload superclass function, use interface, etc.

---

With Dave's provided MySQL dump, I did below items in TablePlus manually before retrieving submissions:
1. Import farms table from live db (to cater farm ID in submission)
2. Update locations table, change records ID from 1-5 to 241-245, to match with records in live DB (to cater location ID in submission)
3. For performance_xxx tables, change performance_assessment_id to nullable

---

TODO:
 - Remove extra double quotes in farms.identifiers JSON value
 - For performance_xxx tables, remove performance_assessment_id, add main_survey_id as foreign key

---

Screen shot:

image

File: src/Services/OdkLinkService.php [hunk 1]
```diff
@@ -726,35 +729,201 @@ private function storeMainSurveyToCustomTable(Xlsform $xlsform, $entry, XlsformT
...
+        // store foreign key column name and foreign key table name in associative array
+        // TODO: find Laravel array helper function to do the same in a simpler way
+        $foreignKeyColumnNames = [];
+        foreach ($foreignKeyDetails as $foreignKey) {
+            foreach ($foreignKey['columns'] as $foreignKeyColumn) {
+                // logger('foreign_key_column_name: ' . $foreignKeyColumn);
+                $foreignKeyColumnNames[$foreignKeyColumn] = $foreignKey['foreign_table'];
+            }
+            // logger('foreign_table: ' . $foreignKey['foreign_table']);
+        }
```
Comment (dave-mills) [inline src/Services/OdkLinkService.php:778]: I like the idea of checking for foreign keys on the database. That seems easier than checking for relationships on the data model, which was my first thought.

**Review by dave-mills — COMMENTED** [review]
I really feel this whole piece about handling farms shouldn't go here. Right now it's *partially* generalised, which is sort of worse than fully generalised or fully specific / custom to the app. 

I think there are a bunch of good ideas in here for when we do come to the generalised version (e.g. getting the list of possible relationships by checking the foreign key) but I don't think we should merge this in. Instead, we should put the farm handling into the main app, and revert this back to the generalised handling of GPS.

**Comment by dan-tang-ssd** [comment]
As discussed, we will not merge this PR into dev branch.
In main repo, we will refer to branch "handle-gps" to use it temporary.

We will need to generalise how to handle foreign key column in a better way.

**Pull request merged**